# Jakub Kosterna
## Allegro Summer e-Xperience 2021  - recruitment task

<h3 id="tocheading">Table of Contents</h3>
<p>
<div id="toc"></div>

### Data & packages load

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import os

import pandas as pd
import numpy as np

from math import sin, cos, sqrt, atan2, radians

import random
np.random.seed(1613)
random.seed(1613)

In [2]:
listOfFile = os.listdir("data/")
allFileNames = list()

for entry in listOfFile:
    fullPath = os.path.join("data/", entry)
    if os.path.isdir(fullPath):
        allFileNames += getListOfFiles(fullPath)
    else:
        allFileNames.append(fullPath)
        
allFileNames = [i.split('/', 1)[-1] for i in allFileNames]
allFileNames = [i.split('.', 1)[0] for i in allFileNames]
allFileNames

['olist_customers_dataset',
 'olist_geolocation_dataset',
 'olist_orders_dataset',
 'olist_order_items_dataset',
 'olist_order_payments_dataset',
 'olist_order_reviews_dataset',
 'olist_products_dataset',
 'olist_sellers_dataset',
 'product_category_name_translation']

In [3]:
dfs = dict()
for fileName in allFileNames:
    df = pd.read_csv("data/" + fileName + ".csv")
    dfs[fileName] = df
    print(fileName, "[", "shape: ", dfs[fileName].shape , "]:", sep = "")
    dfs[fileName].sample(5) 
    #dfs[fileName].describe() # for overview

olist_customers_dataset[shape: (99441, 5)]:


customer_id                customer_unique_id  \
87602  6ca8512df510cd81c156464b01dcd23d  eea6e5f01ce486fd7727e0611cdbaacd   
48546  7df0b66f0f2cba22ba501317ba7c69f9  70601471b692cc0879300373118041ff   
61393  ba76714c4894372325ea2d044823344a  c3cc9ab026e263448a7e804dfada75f5   
18219  cfad1ce34667d10a9c81fc26061f0d6d  49a7d7f106f6adbd40721e2685278da1   
63205  42ce8044da72970c92b48514f86516f5  5842d0c9150cc3e6c4719e97f8d11a73   

       customer_zip_code_prefix       customer_city customer_state  
87602                     84036        ponta grossa             PR  
48546                     37640             extrema             MG  
61393                     24020             niteroi             RJ  
18219                      1526           sao paulo             SP  
63205                     48922  carnaiba do sertao             BA

olist_geolocation_dataset[shape: (1000163, 5)]:


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
147998                         5822       -23.662545       -46.748346   
900969                        88050       -27.541774       -48.524877   
670543                        40213       -12.995025       -38.509865   
491795                        25911       -22.542997       -43.040790   
23868                          2265       -23.470763       -46.585736   

       geolocation_city geolocation_state  
147998        sao paulo                SP  
900969    florianopolis                SC  
670543         salvador                BA  
491795             mage                RJ  
23868         são paulo                SP

olist_orders_dataset[shape: (99441, 8)]:


order_id                       customer_id  \
44357  934f5cbffbb010f2c0b4a1663d2c14fd  0111774394b7b2cd9651eedc6c791bd2   
11578  b148b0338ed3231c5dcc4564f81a7586  7dd168069dddee1868d6293874dbe38b   
46929  49e46b4388f4b9ea96901467bd26eb6d  788cffe37a94186e4ad9a9940f0f858f   
83690  91fb22794b4915b6dddd8e74705905f6  a2e2d0962cced12845a782c6727a6919   
93317  7ba54c67fabcf8bc22e2023e8d81dd84  3facb23e1ea128b605bc112b42adc208   

      order_status order_purchase_timestamp    order_approved_at  \
44357    delivered      2017-07-29 19:21:08  2017-07-29 19:35:12   
11578    delivered      2018-05-11 14:43:47  2018-05-15 05:32:35   
46929    delivered      2018-08-03 20:20:47  2018-08-03 20:30:31   
83690    delivered      2018-01-23 14:55:31  2018-01-23 15:12:46   
93317    delivered      2018-07-19 13:41:59  2018-07-19 13:55:16   

      order_delivered_carrier_date order_delivered_customer_date  \
44357          2017-07-31 15:53:31           2017-08-10 13:53:29   
11578          2018-05-15 11:28:00           2018-05-21 12:42:32   
46929          2018-08-07 14:21:00           2018-08-13 09:47:40   
83690          2018-01-29 18:52:02           2018-02-06 22:04:53   
93317          2018-07-20 15:11:00           2018-07-26 20:02:39   

      order_estimated_delivery_date  
44357           2017-08-24 00:00:00  
11578           2018-05-28 00:00:00  
46929           2018-08-14 00:00:00  
83690           2018-03-05 00:00:00  
93317           2018-08-10 00:00:00

olist_order_items_dataset[shape: (112650, 7)]:


order_id  order_item_id  \
4199   0977997b1c07196d7d76a0d79c50117d              1   
36202  5212d3692378658f5807cd76ed591b13              1   
20576  2f29cfa739331301e37ea8e0cc5c863f              1   
34944  4f19c6b4e6eec789ace7daa84889dd37              2   
80018  b601ce5415b4d58823d60e0e8e32ea12              1   

                             product_id                         seller_id  \
4199   17e7128a33dae59e32be411b7b4921b5  e7f3bf1ecd8526eb7f3b18059e6716e5   
36202  ae7893ab2f9eec6630b5098e6362adc2  41b39e28db005d9731d9d485a83b4c38   
20576  7f4e034245e45549110139d1de248f83  4b9750c8ad28220fe6702d4ecb7c898f   
34944  3581206bcfdc0f2611c3d1e8edb87719  adcf50477d6a1f5d91bb9d54c4903cb5   
80018  ee8b16196604ed47b34b094351ce68e2  1d8dbc4f32378d715c717c1c1fc57bae   

       shipping_limit_date   price  freight_value  
4199   2017-03-28 18:42:09   19.99          10.96  
36202  2017-05-11 13:21:48   14.70           8.72  
20576  2018-02-25 23:30:27   48.90           9.74  
34944  2017-06-12 02:05:15   11.70          15.10  
80018  2017-05-15 09:15:11  109.90          18.94

olist_order_payments_dataset[shape: (103886, 5)]:


order_id  payment_sequential payment_type  \
69457  8580c2f1abe991d5bf205de482913d3c                   1  credit_card   
64592  912fac153319e5a8524a1a9646dcc707                   1       boleto   
21228  3d5f593a3675d379285142d8c932bb6e                   1  credit_card   
49484  36b8ca63eb9822b5f68630749ff70e32                   1  credit_card   
46691  3ddb908d6c179cf271282116f4c5011d                   1       boleto   

       payment_installments  payment_value  
69457                    10         307.27  
64592                     1         248.10  
21228                     7          70.91  
49484                     5          54.33  
46691                     1         148.65

olist_order_reviews_dataset[shape: (100000, 7)]:


review_id                          order_id  \
14872  09359f15483ff74b39b195e6863ce370  cd74303942bee59edc3379f0a6433c62   
56898  e8eee8baab999e4b5ac1a097d1fd6f51  2d53cf390b252a89d9c7fdd57036d7b2   
24355  8d2dcf17b3679ab1484d7d5c65b9507e  1f5bb5f84797dd4e234479124128b9cc   
1627   69be312298c8594fa3e2b8eae718b771  ba227a95d8ae1abc30bda3c6c5c08c1a   
556    d10aa47a4914ed4e55ee13dc41198aa6  95d8d3c4c0f3ac475f40696807344776   

       review_score review_comment_title review_comment_message  \
14872             5                  NaN   foi rápido recomendo   
56898             1                  NaN                    NaN   
24355             5                  NaN                    NaN   
1627              5                  NaN                    NaN   
556               4                  NaN                    NaN   

      review_creation_date review_answer_timestamp  
14872  2018-04-25 00:00:00     2018-04-30 22:11:11  
56898  2017-12-18 00:00:00     2017-12-21 13:53:22  
24355  2017-12-27 00:00:00     2017-12-27 21:40:00  
1627   2018-06-27 00:00:00     2018-06-28 09:50:09  
556    2018-07-12 00:00:00     2018-07-12 19:44:42

olist_products_dataset[shape: (32951, 9)]:


product_id product_category_name  \
16313  557d306a56fe0848efbea81dafc78006            automotivo   
29293  8aa2ca7004be44cabe9d19d3e3763200       cama_mesa_banho   
22305  e3dc529631d8c8db3d80c26b6e05e379       casa_construcao   
4334   3f5a7aa9c9cd21cbfe37688ed326222b            brinquedos   
30309  9baa9ef12ae5a0abd8b741d7ab5bf32d        consoles_games   

       product_name_lenght  product_description_lenght  product_photos_qty  \
16313                 40.0                       906.0                 3.0   
29293                 48.0                       110.0                 2.0   
22305                 58.0                      1547.0                 1.0   
4334                  48.0                       236.0                 1.0   
30309                 43.0                      1078.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
16313           16200.0               56.0               15.0   
29293             300.0               20.0                5.0   
22305           10000.0               60.0               25.0   
4334              600.0               45.0               15.0   
30309             400.0               21.0               18.0   

       product_width_cm  
16313              15.0  
29293              20.0  
22305              25.0  
4334               35.0  
30309              17.0

olist_sellers_dataset[shape: (3095, 4)]:


seller_id  seller_zip_code_prefix  seller_city  \
359   822b63912576852aea9a8436d72317b7                   13840   mogi guacu   
3078  63704069d9bd3a75c1cf59babe56004a                   13212      jundiai   
1868  d13e50eaa47b4cbe9eb81465865d8cfc                    9210  santo andre   
2594  258447aad3f7b01e2643c89f64dc5647                   87210     cianorte   
391   0f0c984a4a955310903b6b6008660bad                   14403       franca   

     seller_state  
359            SP  
3078           SP  
1868           SP  
2594           PR  
391            SP

product_category_name_translation[shape: (71, 2)]:


product_category_name  \
46                                           artes   
18                                 eletroportateis   
13                        tablets_impressao_imagem   
26  moveis_cozinha_area_de_servico_jantar_e_jardim   
52                         sinalizacao_e_seguranca   

              product_category_name_english  
46                                      art  
18                         small_appliances  
13                   tablets_printing_image  
26  kitchen_dining_laundry_garden_furniture  
52                   signaling_and_security

### Joining tables

After reading the above view and the documentation of the data set, I found it convenient to create one new super-table - it will contain information about orders and important information related to them.

Such a solution will not be the best in terms of memory and time, but it will be convenient for quick implementation, clear viewing and my saved time :) Of course, it would be better to stay in the model of several tables commercially.

Frankly speaking not sure if it really won't be too big... if it will be; then I will think about something different ;)

#### Orders + reviews

In [4]:
dfMain = dfs['olist_orders_dataset'].copy()
dfMain = dfMain.drop(["order_status", "order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date",
                      "order_delivered_customer_date", "order_estimated_delivery_date"], axis=1)
dfMain.sample(5)
dfMain.shape

order_id                       customer_id
36542  33632083b7cada8ccfafd21c1a324a46  ec65a85f4c557054399874fb69983469
86625  db4cb930416730b7d23d9a87bc67b497  e59f29432d7235645e652d594ded849b
66746  f2a9a390ab5775c0fddbb8e49ae95614  78f2dde0b75935b85e0d45ba45ff23cd
30939  0b4e790fdbd7616b010599cc3e5f873f  e1fb3bd29f0cc22b821938d2853e4cd5
43092  938e650cbf8099b9464bdfb48c457129  2872fca71254241e3d2879ab4f37ba01

(99441, 2)

I'm going to calculate mean review score for every order.

In [5]:
dfMain = dfMain.merge(dfs['olist_order_reviews_dataset'], on="order_id", how="left")
dfMain = dfMain.drop(["review_id", "review_comment_title", "review_comment_message",
                      "review_creation_date", "review_answer_timestamp"], axis=1)
dfMain.sort_values("order_id").head()
dfMain.shape

order_id                       customer_id  \
85748  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
72250  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
6331   000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
22683  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
5271   00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

       review_score  
85748             5  
72250             4  
6331              5  
22683             4  
5271              5

(100000, 3)

In [6]:
dfMain = dfMain.groupby("order_id").mean("review_score").reset_index()
dfMain.isnull().values.any()
dfMain.groupby("review_score").count().reset_index()

False

review_score  order_id
0       1.000000     11748
1       1.500000         8
2       2.000000      3207
3       2.500000        37
4       3.000000      8241
5       3.333333         1
6       3.500000        26
7       4.000000     19073
8       4.333333         1
9       4.500000        55
10      5.000000     57044

Interesting!

- Each product has been reviewed at least once
- The averages range from 1.0 to 5.0. From the table above it can be concluded that almost every order has been reviewed at most a few times (and of course, due to the shapes, the vast majority only once)

As we can see, most orders have an average grade of at least 4.5. Let me Filter out those with a lower rating - we don't want to recommend those that customers aren't sure about!

In [7]:
dfMain = dfMain[dfMain.review_score >= 4.5]
dfMain.columns = ["order_id", "mean_score"]
dfMain.sample(5)
dfMain.shape

order_id  mean_score
76832  c5c1bc50dd24bd493ca543401aa6d592         5.0
29188  4b019b8fe8e95fc64a3cf3793b35b852         5.0
70598  b60cab8c479e1e80420cbe584d7aa9ac         5.0
17578  2d9d63715c81835d6114d3f608fd9f1c         5.0
73238  bc902a69924e1523c0d7dd83bc7bde91         5.0

(57099, 2)

#### Orders + order items

In [8]:
dfMain = dfMain.merge(dfs['olist_order_items_dataset'], on="order_id")
dfMain = dfMain.drop(["shipping_limit_date", "freight_value", "order_item_id"], axis=1)
dfMain.sample(5)
dfMain.shape

order_id  mean_score  \
24145  6203b82391cd0a3daf1c4dbe763a318f         5.0   
45512  b8d8e7c7fe850dd45f9e3762c8b4b940         5.0   
18516  4b160ae84bf86ca95ce56c7c83c0c5eb         5.0   
13428  36e2905ab14ea657095b38b36727ad07         5.0   
36691  96579b3ca52d82b041b4fc2be5ffb7a9         5.0   

                             product_id                         seller_id  \
24145  130482add9fd75ccb6c57ba007694a2d  87142160b41353c4e5fca2360caf6f92   
45512  4fe43f61e04cd41cadc3bebf606fe77b  4a3ccda38b2129705f3fb522db62ca31   
18516  b40fe421bacf70aea142a25d3ac7737b  e0a366315b1b726b6c7832d664c0f530   
13428  2c42bf0cba650bd08bfeac2412b06ef6  2d50d6282f8aa2257819a77bfaa0efe0   
36691  5c973b4202aac5c372a483e712306f4f  391fc6631aebcf3004804e51b40bcf1e   

        price  
24145  105.00  
45512  139.79  
18516  103.90  
13428   49.00  
36691   49.95

(63249, 5)

#### Orders + sellers

In [9]:
dfMain = dfMain.merge(dfs["olist_sellers_dataset"], on="seller_id")
dfMain.sample(5)
dfMain.shape

order_id  mean_score  \
42630  f8bae43b5bec9eb0d3ae12b8d3b72884         5.0   
11564  643916742191b1ca29968acfeccf30ed         5.0   
44495  fb30f702a40235635b55552094ce6f88         5.0   
7449   5225e054c102b5a1d35f4741e8d6c718         5.0   
7500   6c53d99b0ba56006da10a8213850e9ac         5.0   

                             product_id                         seller_id  \
42630  4082e07ccc62fff9dfaf275a81804952  76c47a299e4a8fe992316a85865acbe9   
11564  8c5876b1c7768217964f353bc7e64393  0db783cfcd3b73998abc6e10e59a102f   
44495  9bc56dc1c13a3dd110faa962cabf1531  2528513dd95219a6013d4d05176e391a   
7449   ab169b33d13e5041b9b9d4a2fafce9d9  ea8482cd71df3c1969d7b9473ff13abc   
7500   6d7a4492be45dfb8e38e762efe6614a4  ea8482cd71df3c1969d7b9473ff13abc   

        price  seller_zip_code_prefix     seller_city seller_state  
42630  199.50                   14075  ribeirao preto           SP  
11564   49.00                   11010          santos           SP  
44495  179.00                    6060          osasco           SP  
7449    24.99                    4160       sao paulo           SP  
7500    27.99                    4160       sao paulo           SP

(63249, 8)

#### Orders + geolocation

Our function will not take specific graphic coordinates, and limiting ourselves to some kind of state measures will practically simplify the subject, while giving the full relevance and usefulness of the information.

In [10]:
geolocation_edited = dfs["olist_geolocation_dataset"].copy()
geolocation_edited.sort_values("geolocation_zip_code_prefix").head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
1246                         1001       -23.549292       -46.633559   
429                          1001       -23.550498       -46.634338   
1182                         1001       -23.549779       -46.633957   
1435                         1001       -23.549292       -46.633559   
326                          1001       -23.551427       -46.634074   

     geolocation_city geolocation_state  
1246        sao paulo                SP  
429         sao paulo                SP  
1182        sao paulo                SP  
1435        sao paulo                SP  
326         sao paulo                SP

For every zip code prefix I'll calculate mean longitude and latitude.

The advantage of this solution is the fact that the average latitude and longitude for each zip code contained in the database will naturally belong to its interior (unless we have contact with some strange non-convex figures), and in cases more crowded by customers - nice estimate the proper "center of life" of the site.

In [11]:
geolocation_edited = geolocation_edited.groupby(["geolocation_zip_code_prefix", "geolocation_city",
                            "geolocation_state"]).mean().reset_index()
geolocation_edited.head()

geolocation_zip_code_prefix geolocation_city geolocation_state  \
0                         1001        sao paulo                SP   
1                         1001        são paulo                SP   
2                         1002        sao paulo                SP   
3                         1002        são paulo                SP   
4                         1003        sao paulo                SP   

   geolocation_lat  geolocation_lng  
0       -23.550215       -46.634019  
1       -23.549998       -46.634060  
2       -23.548438       -46.635129  
3       -23.544641       -46.633180  
4       -23.548988       -46.635786

Ugh! It looks like some cities have been written in different forms. We can handle it as follows:

In [12]:
geolocation_tmp = geolocation_edited.drop(["geolocation_city"], axis=1)
geolocation_tmp = geolocation_tmp.groupby(["geolocation_zip_code_prefix",
                                           "geolocation_state"]).mean().reset_index()
geolocation_tmp.head()

geolocation_zip_code_prefix geolocation_state  geolocation_lat  \
0                         1001                SP       -23.550106   
1                         1002                SP       -23.546540   
2                         1003                SP       -23.549036   
3                         1004                SP       -23.549660   
4                         1005                SP       -23.549650   

   geolocation_lng  
0       -46.634039  
1       -46.634154  
2       -46.635325  
3       -46.634600  
4       -46.635971

In [13]:
geolocation_edited = geolocation_tmp.merge(geolocation_edited.drop(["geolocation_lat", "geolocation_lng"],
                                                                    axis=1), on=["geolocation_zip_code_prefix",
                                                                                 "geolocation_state"])
geolocation_edited.head()
geolocation_edited.shape

geolocation_zip_code_prefix geolocation_state  geolocation_lat  \
0                         1001                SP       -23.550106   
1                         1001                SP       -23.550106   
2                         1002                SP       -23.546540   
3                         1002                SP       -23.546540   
4                         1003                SP       -23.549036   

   geolocation_lng geolocation_city  
0       -46.634039        sao paulo  
1       -46.634039        são paulo  
2       -46.634154        sao paulo  
3       -46.634154        são paulo  
4       -46.635325        sao paulo

(27912, 5)

In [14]:
geolocation_unique_zip_codes = geolocation_edited.groupby(["geolocation_zip_code_prefix", "geolocation_state",
                                                "geolocation_lat", "geolocation_lng"]).min().reset_index()
geolocation_unique_zip_codes.head()
geolocation_unique_zip_codes.shape

geolocation_zip_code_prefix geolocation_state  geolocation_lat  \
0                         1001                SP       -23.550106   
1                         1002                SP       -23.546540   
2                         1003                SP       -23.549036   
3                         1004                SP       -23.549660   
4                         1005                SP       -23.549650   

   geolocation_lng geolocation_city  
0       -46.634039        sao paulo  
1       -46.634154        sao paulo  
2       -46.635325        sao paulo  
3       -46.634600        sao paulo  
4       -46.635971        sao paulo

(19023, 5)

Finally! Now we have exactly one location for each zip code in the data.

Let's merge it with the main data frame.

In [15]:
dfMain = dfMain.drop(["seller_city"], axis=1)
dfMain = dfMain.merge(geolocation_unique_zip_codes,
                       left_on=["seller_zip_code_prefix", "seller_state"],
                       right_on=["geolocation_zip_code_prefix", "geolocation_state"])
dfMain = dfMain.drop(["geolocation_zip_code_prefix", "geolocation_state"], axis=1)
dfMain.columns = ["order_id", "mean_score", "product_id", "seller_id", "price",
                   "zip_code_prefix", "state", "lat", "lng", "city"]
dfMain = dfMain[["order_id", "product_id", "seller_id", "mean_score", "state",
                   "city", "zip_code_prefix", "lat", "lng"]]
dfMain.sample(5)
dfMain.shape

order_id                        product_id  \
10864  3e940e5beb328b427d180e7e8165bafb  efd7af8312533d58ccb5ac26304dd3d5   
54423  b30b24f06aaf93a002acff534e7bddff  629576b997e202bfb375fc3538a1c838   
32232  e43f88863d1cc9cff18d753d439d42e8  2e0b9273624db778d82fd3b18becec24   
13817  638dd0cfabf15189b7ce42b077f9fb9d  53ea9da485f6aed8a6f03a85831fe021   
36306  70f814478ea3019fff8677f0aca6a54b  156bb7508d25ad3108035c7ec3a3d2fb   

                              seller_id  mean_score state       city  \
10864  ea8482cd71df3c1969d7b9473ff13abc         5.0    SP  sao paulo   
54423  abcd2cb37d46c2c8fb1bf071c859fc5b         5.0    MT     cuiaba   
32232  de23c3b98a88888289c6f5cc1209054a         5.0    SP  sao paulo   
13817  7e93a43ef30c4f03f38b393420bc753a         5.0    SP    barueri   
36306  77a515caa36327151d1cc6c32a9f00e1         5.0    SP    diadema   

       zip_code_prefix        lat        lng  
10864             4160 -23.625520 -46.611898  
54423            78020 -15.602953 -56.102606  
32232             5530 -23.576961 -46.717157  
13817             6429 -23.489688 -46.952915  
36306             9912 -23.689613 -46.615598

(62709, 9)

We lost some data, but not so much. Probably some rows having zip codes present in the main dataset, but not in the geolocation info one. Also - maybe - a few zip codes being on the border between states? Either way, it's a very small percentage of all observations - it shouldn't have a significant impact on the algorithm. The coordinates are not in the first place anyway and are not taken into account by the function.

#### Orders + products

In [16]:
dfMain = dfMain.merge(dfs["olist_products_dataset"], on="product_id")
dfMain.sample(5)
dfMain.shape

order_id                        product_id  \
44824  9d112e41d85aec5101964866c4801784  48c5608bfb241a35beb7a00d960cd27f   
57055  2d3518f1b535de87ccf593d3892022af  c319f453652bff80bc8f9f71d87e56c1   
44981  03faedb6f3c4299c7cd934f5242b194c  caf230c721f149691d0307f0868beac6   
18880  dd082c5f38a7001c40cf57a14f1b3b34  02ee0862685e90b57cc241b7c9885f95   
44274  1ab14cf1dddb5af11ba2016a5368c210  e19c52ba2deb45c56077643f204eda1e   

                              seller_id  mean_score state  \
44824  c68fb906c8f4b4b946d8386bfa6e5467         5.0    SP   
57055  f680f85bee2d253556ac91be391d2c82         5.0    SP   
44981  b4ffb71f0cb1b1c3d63fad021ecf93e1         5.0    SP   
18880  fe2032dab1a61af8794248c8196565c9         5.0    SP   
44274  b372ee768ed69e46ca8cdbd267aa7a38         5.0    SP   

                        city  zip_code_prefix        lat        lng  \
44824            jaboticabal            14870 -21.257174 -48.320964   
57055            carapicuiba             6365 -23.552502 -46.851300   
44981              sao paulo             3880 -23.508364 -46.485316   
18880               campinas            13030 -22.924970 -47.074284   
44274  sao jose do rio preto            15013 -20.851542 -49.259116   

             product_category_name  product_name_lenght  \
44824        utilidades_domesticas                 46.0   
57055                        bebes                 46.0   
44981                   automotivo                 59.0   
18880                   perfumaria                 55.0   
44274  fashion_bolsas_e_acessorios                 49.0   

       product_description_lenght  product_photos_qty  product_weight_g  \
44824                       509.0                 1.0             700.0   
57055                       715.0                 6.0             260.0   
44981                       512.0                 2.0             500.0   
18880                       784.0                 1.0             555.0   
44274                       566.0                 1.0             100.0   

       product_length_cm  product_height_cm  product_width_cm  
44824               29.0               13.0              13.0  
57055               24.0               19.0              20.0  
44981               30.0               20.0              20.0  
18880               24.0                8.0              15.0  
44274               22.0                2.0              19.0

(62709, 17)

Ok, it's definitely not too huge!

### Algorithm for logged in user

We won't use most of *dfMain* columns actually.

In [17]:
dfLogged = dfMain[["product_id", "state", "city", "zip_code_prefix", "lat", "lng", "product_category_name"]]
dfLogged.sample(5)

product_id state                   city  \
44290  1ba74b19c790ee59fb57fa2befa6e885    SP  sao jose do rio preto   
43165  f35927953ed82e19d06ad3aac2f06353    SC             rio do sul   
1732   35afc973633aaeb6b877ff57b2793310    SP               ibitinga   
23201  45e967683e7292b195609137fadaf2fe    SP              bebedouro   
14627  fc935e2845dacecf29f0555465bed91b    SP                barueri   

       zip_code_prefix        lat        lng        product_category_name  
44290            15013 -20.851542 -49.259116  fashion_bolsas_e_acessorios  
43165            89160 -27.214587 -49.637725       livros_interesse_geral  
1732             14940 -21.757321 -48.829744                casa_conforto  
23201            14709 -20.924600 -48.490591                   automotivo  
14627             6429 -23.489688 -46.952915           relogios_presentes

The following symbols - as in the document.

#### A

In [18]:
def findA(df, favCats, location):
    df = df[df["city"] == location[0]]
    df = df.loc[df['product_category_name'].isin(favCats)]
    df = df.groupby(["product_id"]).count().reset_index()
    result_row = df.loc[df['city'].idxmax()]
    return result_row.product_id

Example usage:

In [19]:
findA(dfLogged, ["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

'8aa6223e400af9c97b07c75993142721'

#### B

In order to calculate it, we need function for finding 5 nearest cities, in which at least one product of verified category was sold.

In [20]:
geolocation_unique_cities = geolocation_unique_zip_codes.drop(["geolocation_zip_code_prefix"], axis=1)
geolocation_unique_cities = geolocation_unique_cities.groupby(["geolocation_state",
                                                               "geolocation_city"]).mean().reset_index()
geolocation_unique_cities.columns = ["state", "city", "lat", "lng"]
geolocation_unique_cities.sample(5)

state                 city        lat        lng
3195    PE            ibirajuba  -8.580398 -36.178220
958     ES             paulista -18.664810 -40.794957
96      AL  passo de camaragibe  -9.250592 -35.486157
1123    GO               itaucu -16.202128 -49.608312
2763    MT                 vera -12.288099 -55.296991

In [21]:
def distance(lat1, lon1, lat2, lon2):
    R = 6373.0 # approximate radius of earth in km
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return R * c

Example usage:

In [22]:
print(distance(52.2296756, 21.0122287, 52.406374, 16.9251681))

5447.055461466734


In [23]:
def getCitiesContainingCategorySell(dfProducts, categoryName):
    df_cities_category_sold = dfProducts[dfProducts["product_category_name"] == categoryName]
    cities_selling = df_cities_category_sold[["city"]].drop_duplicates()
    return list(cities_selling["city"])

Example usage:

In [24]:
getCitiesContainingCategorySell(dfLogged, "utilidades_domesticas")[0:10]
len(getCitiesContainingCategorySell(dfLogged, "utilidades_domesticas"))

['loanda',
 'ibitinga',
 'sao jose do rio preto',
 'cascavel',
 'maua',
 'monte siao',
 'guarulhos',
 'piracicaba',
 'sao paulo',
 'araguari']

145

In [25]:
def find5nearest(dfProducts, dfLocations, cityName, categoryName):
    cityRow = dfLocations[dfLocations['city'] == cityName].iloc[0]
    ourLat = cityRow.lat
    ourLng = cityRow.lng
    
    citiesContainingCat = getCitiesContainingCategorySell(dfProducts, categoryName)
    dfLocations = dfLocations[dfLocations["city"].isin(citiesContainingCat)]
    
    dfLocations["distance"] = dfLocations.apply(lambda x: distance(ourLat, ourLng, x["lat"], x["lng"]), axis=1)
    dfLocations = dfLocations.sort_values("distance")
    dfLocations = dfLocations[dfLocations["distance"] > 0]
    dfLocations = dfLocations.head(5)
    return list(dfLocations["city"])

Example usage:

In [26]:
pd.options.mode.chained_assignment = None

find5nearest(dfLogged, geolocation_unique_cities, "guarulhos", "utilidades_domesticas")[0:10]
len(find5nearest(dfLogged, geolocation_unique_cities, "guarulhos", "utilidades_domesticas"))

['aruja', 'barueri', 'osasco', 'mogi das cruzes', 'sorocaba']

5

In [27]:
def findB(df, favCats, location):
    result = list()
    for cat in favCats:
        nearestCities = find5nearest(dfLogged, geolocation_unique_cities, location[0], cat)
        dfTmp = df[df["state"] == location[1]]
        dfTmp = df[~df["city"].isin(nearestCities)]
        dfTmp = dfTmp.loc[dfTmp['product_category_name'] == cat]
        dfTmp = dfTmp.groupby(["product_id"]).count().reset_index()
        result_row = dfTmp.loc[dfTmp['city'].idxmax()]
        result.append(result_row.product_id)
    return result

Example usage:

In [28]:
findB(dfLogged, ["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

['c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1']

#### C

In [29]:
def findC(df, favCats, location):
    result = list()
    for cat in favCats:
        nearestCities = find5nearest(dfLogged, geolocation_unique_cities, location[0], cat)
        dfTmp = df[df["city"].isin(nearestCities)]
        dfTmp = dfTmp.loc[dfTmp['product_category_name'] == cat]
        dfTmp = dfTmp.groupby(["product_id"]).count().reset_index()
        result_row = dfTmp.loc[dfTmp['city'].idxmax()]
        result.append(result_row.product_id)
    return result

In [30]:
findC(dfLogged, ["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

['008cff0e5792219fae03e570f980b330',
 '306e6b5dda3397e4a9312f94d82565b5',
 '3ce21e38e6a3060c20f4e74bdab770c8']

#### D

In [31]:
def findD(df, favCats, location):
    anyNone = False
    result = list()
    for cat in favCats:
        dfTmp = df[df["city"] == location[0]]
        dfTmp = df[df["product_category_name"] == cat]
        if len(dfTmp) > 0:
            dfTmp = dfTmp.groupby(["product_id"]).count().reset_index()
            result_row = dfTmp.loc[dfTmp['city'].idxmax()]
            result.append(result_row.product_id)
        else:
            dfTmp = df.copy()
            if not anyNone:
                already_chosen = findC(df, favCats, location)
            else:
                anyNone = True
            dfTmp = dfTmp[~dfTmp["product_id"].isin(already_chosen)]
            nearestCities = find5nearest(dfLogged, geolocation_unique_cities, location[0], cat)
            dfTmp = dfTmp[dfTmp["city"].isin(nearestCities)]
            dfTmp = dfTmp.loc[dfTmp['product_category_name'] == cat]
            dfTmp = dfTmp.groupby(["product_id"]).count().reset_index()
            result_row = dfTmp.loc[dfTmp['city'].idxmax()]
            result.append(result_row.product_id)
    return result

Example usage:

In [32]:
findD(dfLogged, ["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

['c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1']

### Algorithm for not logged in user

#### First step

Having almost zero information, we have a large pool of products to choose from. Let us limit these to further analysis only to those with the maximum average rating - all 5.0. It will also be displayed nicely and will inspire you in a positive way, encouraging you to buy even not necessarily including these products.

In [33]:
dfMain2 = dfMain[dfMain["mean_score"] == 5.0]
dfMain2.sample(5)
dfMain2.shape

order_id                        product_id  \
52677  6976421e858af2e4b388ccfac68236ee  9246752373bd625f47da5e3d8b41b1dc   
47974  8d282c2ddced263b2c6ddf398e3fb6dc  2280139813948df83ba872aecca72997   
42004  6d3387429df1d01b1fd17ffe1b631d31  1bfe10e822dedba7190af59828c822d7   
13153  805559bf950cf417d71f74d0d92b4302  593236d0ff46b4299b4787fb8d43f7f0   
12471  0473cdb86b1d6469287e484366a1a608  d7423500e1911ba5972c229b2fb974ca   

                              seller_id  mean_score state  \
52677  4a82c4af97ffc0fb2dc26bfdc03b1842         5.0    MG   
47974  2528513dd95219a6013d4d05176e391a         5.0    SP   
42004  f5a590cf36251cf1162ea35bef76fe84         5.0    SP   
13153  0dd184061fb0eaa7ca37932c68ab91c5         5.0    SP   
12471  7299e27ed73d2ad986de7f7c77d919fa         5.0    MG   

                        city  zip_code_prefix        lat        lng  \
52677         lagoa da prata            35590 -20.024065 -45.537190   
47974                 osasco             6060 -23.553054 -46.781981   
42004  sao bernardo do campo             9720 -23.715758 -46.553994   
13153              guarulhos             7031 -23.493719 -46.546189   
12471               araguari            38440 -18.652644 -48.192235   

               product_category_name  product_name_lenght  \
52677                            NaN                  NaN   
47974                  casa_conforto                 60.0   
42004          instrumentos_musicais                 60.0   
13153  industria_comercio_e_negocios                 44.0   
12471                 consoles_games                 25.0   

       product_description_lenght  product_photos_qty  product_weight_g  \
52677                         NaN                 NaN             700.0   
47974                       249.0                 3.0            3500.0   
42004                       814.0                 3.0            9150.0   
13153                      1396.0                 2.0            7225.0   
12471                      2247.0                 1.0             175.0   

       product_length_cm  product_height_cm  product_width_cm  
52677               20.0               21.0              20.0  
47974               40.0               20.0              40.0  
42004               50.0               50.0              50.0  
13153               16.0              103.0              25.0  
12471               20.0                2.0              15.0

(62640, 17)

#### Second step

Let's merge it again with *reviews* - to filter only these having some nice comments.

In [34]:
dfMain2 = dfMain2.merge(dfs['olist_order_reviews_dataset'], on="order_id", how="left")
dfMain2 = dfMain2.drop(["review_id", "review_score", "review_comment_title",
                        "review_creation_date", "review_answer_timestamp"], axis=1)
dfMain2 = dfMain2[dfMain2['review_comment_message'].notnull()]
dfMain2 = dfMain2[dfMain2['review_comment_message'].str.len() >= 20]
dfMain2 = dfMain2[dfMain2['review_comment_message'].str.len() <= 50]
dfMain2 = dfMain2.dropna()
dfMain2.sample(5)
dfMain2.shape

order_id                        product_id  \
34741  96b7f5a470ca6a58f261be1af286e4d9  ddceb6eef6af67e9fbbd4c148dbe5fd9   
14805  e102f8dfd6da168838a2d13dfbc3bb10  e0d64dcfaa3b6db5c54ca298ae101d05   
30275  f6dbef2d61321dedc716a1683d7b1509  67523567791b73064e78d6a5731f379c   
4979   d4a72eff3bfdb6ff7667bef4119283e9  c0737ec7b87906c79f130f39cec3d3ee   
20162  c36722bf525388104790764a7917ca07  381169576b9f3c083a1c3261da161a3b   

                              seller_id  mean_score state       city  \
34741  ceaec5548eefc6e23e6607c5435102e7         5.0    SP  sao paulo   
14805  4869f7a5dfa277a7dca6462dcf3b52b2         5.0    SP    guariba   
30275  4e922959ae960d389249c378d1c939f5         5.0    SP    jacarei   
4979   7ad32824caee82087b3e2e5f33b1bf32         5.0    SP   ibitinga   
20162  8c16d1f32a54d92897cc437244442e1b         5.0    SC   blumenau   

       zip_code_prefix        lat        lng   product_category_name  \
34741             3821 -23.496558 -46.501630              cool_stuff   
14805            14840 -21.363502 -48.229601      relogios_presentes   
30275            12327 -23.302385 -45.972127              automotivo   
4979             14940 -21.757321 -48.829744         cama_mesa_banho   
20162            89023 -26.949943 -49.076925  informatica_acessorios   

       product_name_lenght  product_description_lenght  product_photos_qty  \
34741                 52.0                       516.0                 1.0   
14805                 35.0                       514.0                 2.0   
30275                 39.0                      2642.0                 2.0   
4979                  29.0                       248.0                 1.0   
20162                 43.0                       677.0                 2.0   

       product_weight_g  product_length_cm  product_height_cm  \
34741            4650.0               37.0               20.0   
14805            4338.0               17.0               12.0   
30275              83.0               19.0                4.0   
4979             1025.0               35.0               10.0   
20162             225.0               25.0                9.0   

       product_width_cm                           review_comment_message  
34741              34.0     Tá de parabéns rapidez na entrega Maravilha   
14805              14.0   Chegou bem antes do prazo estimado, recomendo!  
30275              14.0                             Como sempre ótimos!!  
4979               18.0  Produto de ótima qualidade, conforme anunciado.  
20162              29.0            Parabéns pela presteza no atendimento

(8270, 18)

#### Third step

In [35]:
dfMain2 = dfMain2[dfMain2['product_name_lenght'] >= 30]
dfMain2 = dfMain2[dfMain2['product_name_lenght'] <= 60]
dfMain2.sample(5)
dfMain2.shape

order_id                        product_id  \
23439  6234c0a631c7ef6fa2bf006aadf30830  ecb6e29c8e01d5fbc73da1d0f12c088e   
12140  e3866f0f605208a1b7fc42b0dddc0da6  061b1614b0e2a54bb06f23af8308b55d   
27914  57e4fc8f20507381b3f7abbe4e2ccda3  6ca11e79ed1acb748ccbb8e6b821967d   
28158  3716f40c5211dc25902ffd174d8d198b  13ce216ee959a22dbaec66c182a1b87f   
35529  4a53124ff0021d4eb276c76766d45f71  c64490cdfe40f9e25abd222c541a06f0   

                              seller_id  mean_score state              city  \
23439  ebd2d60905fb58271facef5596b620d3         5.0    SP  lencois paulista   
12140  ea8482cd71df3c1969d7b9473ff13abc         5.0    SP         sao paulo   
27914  17e34d8224d27a541263c4c64b11a56b         5.0    SP    ribeirao preto   
28158  a416b6a846a11724393025641d4edd5e         5.0    SP         sao paulo   
35529  5a8e7d5003a1f221f9e1d6e411de7c23         5.0    SP          campinas   

       zip_code_prefix        lat        lng product_category_name  \
23439            18683 -22.601625 -48.807015          beleza_saude   
12140             4160 -23.625520 -46.611898             telefonia   
27914            14085 -21.165989 -47.798251            perfumaria   
28158             3702 -23.515978 -46.547372          beleza_saude   
35529            13076 -22.881709 -47.050464      moveis_decoracao   

       product_name_lenght  product_description_lenght  product_photos_qty  \
23439                 43.0                      1679.0                 1.0   
12140                 60.0                       625.0                 3.0   
27914                 49.0                       896.0                 2.0   
28158                 52.0                      1376.0                 1.0   
35529                 58.0                      2017.0                 2.0   

       product_weight_g  product_length_cm  product_height_cm  \
23439            2150.0               33.0               20.0   
12140             350.0               17.0                4.0   
27914             400.0               16.0               16.0   
28158             500.0               30.0               10.0   
35529             600.0               68.0                8.0   

       product_width_cm                          review_comment_message  
23439              17.0     ainda não utilizei mas a entrega foi otima   
12140              12.0              PEDIDO DE ACORDO COM O SOLICITADO.  
27914              22.0        Tudo ótimo. Entregaram antes do prazo!!!  
28158              17.0  entrega dentro do prazo corretamente,recomendo  
35529              13.0                          Bom o serviço testado.

(7739, 18)

Btw: why is it *lenght*, not *length*? 😂 Let's do some cleaning.

In [36]:
dfMain2 = dfMain2.drop(["order_id", "seller_id", "zip_code_prefix", "lat", "lng",
                               "product_name_lenght", "product_description_lenght", "review_comment_message"], axis=1)
dfMain2.sample(5)
dfMain2.shape

product_id  mean_score state             city  \
54019  6cbb9a0a2b5565f3b3afb8185f6446f0         5.0    SP        sao paulo   
48567  ce582863190cb932e8a7f5515791c11e         5.0    SP  itaquaquecetuba   
22419  f65da4eafdfa856de29dc7392408dd0a         5.0    SP            salto   
60097  00ae7076313576f94d9107599d79a978         5.0    SP        sao paulo   
34587  64cc3c7698b70c71f7f1ce671c112ad0         5.0    SP           franca   

      product_category_name  product_photos_qty  product_weight_g  \
54019         esporte_lazer                 1.0             300.0   
48567            automotivo                 7.0             550.0   
22419            brinquedos                 3.0            4600.0   
60097         esporte_lazer                 1.0             600.0   
34587          beleza_saude                 2.0             150.0   

       product_length_cm  product_height_cm  product_width_cm  
54019               27.0                9.0              14.0  
48567               30.0                5.0              20.0  
22419               70.0               10.0              38.0  
60097               22.0               33.0              16.0  
34587               20.0               20.0              20.0

(7739, 10)

#### Forth step

In [37]:
dfMain2 = dfMain2[dfMain2["product_photos_qty"] >= 2]
dfMain2 = dfMain2.drop(["product_photos_qty"], axis=1)
dfMain2.sample(5)
dfMain2.shape

product_id  mean_score state  \
5959   389d119b48cf3043d311335e499d9c6b         5.0    SP   
50317  4052517cac9e78357d895976124f6972         5.0    SP   
56465  ce9b5e2f3503d1142f9c9866db452633         5.0    PR   
56643  5026f48060b51abce464cb03e1d92847         5.0    GO   
46449  a8b0dd2c57b49b649373ff5a495a697e         5.0    RJ   

                        city   product_category_name  product_weight_g  \
5959   sao jose do rio preto      ferramentas_jardim            1750.0   
50317              sao paulo             eletronicos             800.0   
56465              apucarana   utilidades_domesticas            1800.0   
56643                goiania                   audio             450.0   
46449            sao goncalo  informatica_acessorios             350.0   

       product_length_cm  product_height_cm  product_width_cm  
5959                37.0               22.0              40.0  
50317               35.0               30.0              40.0  
56465               30.0               50.0              30.0  
56643               20.0                9.0              15.0  
46449               17.0                7.0              13.0

(3885, 9)

#### Five step

Let me take a look at the number of products "left in the pot".

In [38]:
dfMain2countTmp = dfMain2[["product_id", "product_category_name"]].groupby(["product_id"]).count().reset_index()
dfMain2countTmp.sample(5)
dfMain2countTmp.shape

product_id  product_category_name
984   65906875895d4c8ba11779d8bd5b1270                      1
2371  f8f93700c26007f74716eca57d217cfd                      1
2238  ed724a28107bfd2a1aebcb6dcd2055f4                      1
1171  77d1b9fb8d3858f503316105e9ac6c51                      1
124   0dce093fc7ec4dec214256df878e3c08                      1

(2441, 2)

OK! So we have 2441 unique products; about 40% less than earlier. What about products, which were sold more than 5 times?

In [39]:
dfMain2countTmp = dfMain2countTmp[dfMain2countTmp["product_category_name"] >= 5]
dfMain2countTmp.columns = ["product_id", "sales_count"]
dfMain2countTmp.sample(5)
dfMain2countTmp.shape

product_id  sales_count
44    05b4647d7e40f280dc06e3d69deaba21            5
915   5e2ba75ad255ff60b1c76c5bf526ae9b            5
883   5a3320037d5922a7708647c81ecc1f15            5
1587  a62e25e09e05e6faf31d90c6ec1aa3d1           21
730   4bb4fb9c85785b75ab4f6559900c7ca1            5

(93, 2)

Great! Less than 100 products to choose from.

In [40]:
dfMain2 = dfMain2.merge(dfMain2countTmp, on = "product_id")
dfMain2 = dfMain2.drop_duplicates()
dfMain2 = dfMain2.drop(["mean_score"], axis=1)
dfMain2.sample(5)

product_id state                   city  \
570  72172e982e8b92155069e4201c92c0bb    SP           praia grande   
741  9e2d3a8d8ffad53e2e35282a2020221c    MG                    uba   
215  389d119b48cf3043d311335e499d9c6b    SP  sao jose do rio preto   
473  52c80cedd4e90108bf4fa6a206ef6b03    MG               ilicinea   
407  d285360f29ac7fd97640bf0baef03de0    SP               campinas   

    product_category_name  product_weight_g  product_length_cm  \
570         esporte_lazer             200.0               28.0   
741      moveis_decoracao            1600.0               35.0   
215    ferramentas_jardim            1750.0               37.0   
473    ferramentas_jardim            7650.0               20.0   
407    relogios_presentes             363.0               18.0   

     product_height_cm  product_width_cm  sales_count  
570               10.0              13.0            6  
741                6.0              32.0            5  
215               22.0              40.0           46  
473               20.0              20.0           11  
407               13.0              13.0            7

#### Sixth step

In [41]:
dfMain2countTmp = dfMain2[["city", "product_id"]].groupby(["city"]).count().reset_index()
dfMain2countTmp = dfMain2countTmp.sort_values("product_id")
dfMain2countTmp.tail()
dfMain2countTmp.shape

city  product_id
27         ribeirao preto           3
6                campinas           4
34  sao jose do rio preto           6
31            santo andre           8
37              sao paulo          29

(42, 2)

For each city with >= 3 products, from the table I will remove the *floor(1/2)* of the products from that city that sold the least.

In [42]:
popularCities = dfMain2countTmp[dfMain2countTmp["product_id"] >= 3]
popularCities["how_many_to_remove"] = popularCities["product_id"] // 2
popularCities

city  product_id  how_many_to_remove
29                  salto           3                   1
11                guariba           3                   1
19       lauro de freitas           3                   1
18              joinville           3                   1
27         ribeirao preto           3                   1
6                campinas           4                   2
34  sao jose do rio preto           6                   3
31            santo andre           8                   4
37              sao paulo          29                  14

In [43]:
isFirstIteration = True
for index, row in popularCities.iterrows():
    cityHere = row["city"]
    howManyToRemove = row['how_many_to_remove']
    dfMain2tmp = dfMain2[dfMain2["city"] == cityHere]
    dfMain2tmp = dfMain2tmp.sort_values("sales_count")
    dfMain2tmp = dfMain2tmp.iloc[howManyToRemove:]
    if isFirstIteration:
        dfMain2final = dfMain2tmp
        isFirstIteration = False
    else:
        dfMain2final = dfMain2final.append(dfMain2tmp)
dfMain2final.sample(5)
dfMain2final.shape

product_id state                   city  \
868  f288cae3916218ec01f708d093b29d4f    SP              sao paulo   
816  2a2d22ae30e026f1893083c8405ca522    SC              joinville   
490  d017a2151d543a9885604dc62a3d9dcc    SP              sao paulo   
83   368c6c730842d78016ad823897a372db    SP  sao jose do rio preto   
359  97017430754804328eb9597b7f85da03    SP              sao paulo   

           product_category_name  product_weight_g  product_length_cm  \
868           ferramentas_jardim             150.0               20.0   
816              cama_mesa_banho            8350.0               50.0   
490  fashion_bolsas_e_acessorios             250.0               16.0   
83            ferramentas_jardim            1650.0               30.0   
359                    telefonia             300.0               17.0   

     product_height_cm  product_width_cm  sales_count  
868                5.0              15.0            6  
816               27.0              38.0            7  
490                2.0              11.0           12  
83                22.0              30.0           39  
359                4.0              12.0            8

(34, 9)

In [44]:
dfMain2final = dfMain2final.append(dfMain2[~dfMain2["city"].isin(popularCities["city"])])
dfMain2final.sample(5)
dfMain2final.shape

product_id state          city  \
731  23bcd6822a33df5534f9b290216eec1f    RS        canoas   
359  97017430754804328eb9597b7f85da03    SP     sao paulo   
746  f7fce9a4e86a0d19818cc145ef87f8ca    SP       limeira   
570  72172e982e8b92155069e4201c92c0bb    SP  praia grande   
319  aca2eb7d00ea1a7b8ebd4e68314663af    SP     sao paulo   

           product_category_name  product_weight_g  product_length_cm  \
731                   brinquedos            7000.0               20.0   
359                    telefonia             300.0               17.0   
746  fashion_bolsas_e_acessorios             200.0               16.0   
570                esporte_lazer             200.0               28.0   
319             moveis_decoracao            2600.0               50.0   

     product_height_cm  product_width_cm  sales_count  
731               62.0              35.0           10  
359                4.0              12.0            8  
746                2.0              11.0            5  
570               10.0              13.0            6  
319               10.0              30.0           26

(75, 9)

In [45]:
bestProductsNotLoggedIn = list(dfMain2final["product_id"])
bestProductsNotLoggedIn[1:10]
len(bestProductsNotLoggedIn)

['f7a17d2c51d9df89a4f1711c4ac17f33',
 '461f43be3bdf8844e65b62d9ac2c7a5a',
 '53b36df67ebb7c41585e8d54d6772e08',
 '431d674f9a4fbd0957ecf6ba3fcb6899',
 '6ae3f85f46a37844372252496c77b4d8',
 '25f86162fee18735fffdb762dcb10d7c',
 '2a2d22ae30e026f1893083c8405ca522',
 '461f43be3bdf8844e65b62d9ac2c7a5a',
 '53b36df67ebb7c41585e8d54d6772e08']

75

### Side functions

In [46]:
def predictKnown(favCats, location):
    listA = findA(dfLogged, favCats, location)
    result_list = [listA]
    
    listB = findB(dfLogged, favCats, location)
    listC = findC(dfLogged, favCats, location)
    listD = findD(dfLogged, favCats, location)
    BCD = (listB, listC, listD)
    
    for sublist in BCD:
        for item in sublist:
            result_list.append(item)
    
    return result_list

Example usage:

In [47]:
predictKnown(["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

['8aa6223e400af9c97b07c75993142721',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1',
 '008cff0e5792219fae03e570f980b330',
 '306e6b5dda3397e4a9312f94d82565b5',
 '3ce21e38e6a3060c20f4e74bdab770c8',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1']

In [48]:
def predictUnknown():
    return random.sample(bestProductsNotLoggedIn, 10)

In [49]:
predictUnknown()

['461f43be3bdf8844e65b62d9ac2c7a5a',
 '461f43be3bdf8844e65b62d9ac2c7a5a',
 'a62e25e09e05e6faf31d90c6ec1aa3d1',
 'f288cae3916218ec01f708d093b29d4f',
 '52c80cedd4e90108bf4fa6a206ef6b03',
 'fb55982be901439613a95940feefd9ee',
 '53b36df67ebb7c41585e8d54d6772e08',
 '4c2394abfbac7ff59ec7a420918562fa',
 '431d674f9a4fbd0957ecf6ba3fcb6899',
 'd285360f29ac7fd97640bf0baef03de0']

### Final function

In [50]:
def predictTopProducts(favCats = None, location = None):
    if favCats is None or location is None:
        return predictUnknown()
    else:
        try:
            favCats = np.array(favCats)
            location = np.array(location)
        except Exception:
            print("Favorite categories and locations should be objects capable of being dumped into the numpy array")
        else:
            assert isinstance(favCats, np.ndarray)
            assert isinstance(location, np.ndarray)
            for i in range(3):
                assert isinstance(favCats[i], str), "Favorite categories should be strings"
            assert len(favCats) == 3, "Exactly three favorite categories needed"
            assert len(location) == 2, "Exactly two location strings [customer_city & customer_state] needed"
            assert isinstance(location[0], str) and isinstance(location[1], str), "Locations should be strings"
            return predictKnown(favCats, location)

Example usage:

In [51]:
predictTopProducts(["esporte_lazer", "brinquedos", "bebes"], ["tupa", "SP"])

['8aa6223e400af9c97b07c75993142721',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1',
 '008cff0e5792219fae03e570f980b330',
 '306e6b5dda3397e4a9312f94d82565b5',
 '3ce21e38e6a3060c20f4e74bdab770c8',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 '880be32f4db1d9f6e2bec38fb6ac23ab',
 'cac9e5692471a0700418aa3400b9b2b1']

In [52]:
predictTopProducts()

['d285360f29ac7fd97640bf0baef03de0',
 '461f43be3bdf8844e65b62d9ac2c7a5a',
 '53b36df67ebb7c41585e8d54d6772e08',
 '46ca9515def9ceb56e338c86ca765ba1',
 '473795a355d29305c3ea6b156833adf5',
 'bee2e070c39f3dd2f6883a17a5f0da45',
 '7c1bd920dbdf22470b68bde975dd3ccf',
 'f1fe595ee7ef768b41bd9b246d13432d',
 'fb55982be901439613a95940feefd9ee',
 '75d6b6963340c6063f7f4cfcccfe6a30']

### Product lists

#### Case 1

In [53]:
predictTopProducts(["cama_mesa_banho", "papelaria", "fashion_calcados"], ["sao paulo", "SP"])

['f1c7f353075ce59d8a6f3cf58f419c9c',
 '99a4788cb24856965c36a24e339b6058',
 '5411e9269501a870cabf632f05655131',
 'a2c75a23c2f838881dd4275c0cec519f',
 'f1c7f353075ce59d8a6f3cf58f419c9c',
 'e03102efbc2229024c89be731f0aedcb',
 'ac5e164e2eda939ffa46593f90077f9a',
 '99a4788cb24856965c36a24e339b6058',
 '5411e9269501a870cabf632f05655131',
 'ac5e164e2eda939ffa46593f90077f9a']

#### Case 2

In [54]:
predictTopProducts(["esporte_lazer", "moveis_decoracao", "telefonia"], ["rio de janeiro", "RJ"])

['71a5f1c2a5fd9889ef26b5ac22aec9c6',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 'aca2eb7d00ea1a7b8ebd4e68314663af',
 '97017430754804328eb9597b7f85da03',
 '054515fd15bc1a2029f10de97ffa9120',
 '569ffd16f8032478cbeb9800f2e94ba0',
 '8c5876b1c7768217964f353bc7e64393',
 'c6336fa91fbd87c359e44f5dca5a90ed',
 'aca2eb7d00ea1a7b8ebd4e68314663af',
 '97017430754804328eb9597b7f85da03']

#### Case 3

In [55]:
predictTopProducts()

['bee2e070c39f3dd2f6883a17a5f0da45',
 '0aabfb375647d9738ad0f7b4ea3653b1',
 'f1c7f353075ce59d8a6f3cf58f419c9c',
 'c1488892604e4ba5cff5b4eb4d595400',
 'cfd414b4463647f58c7775eaae06893d',
 'd017a2151d543a9885604dc62a3d9dcc',
 '629e019a6f298a83aeecc7877964f935',
 'aca2eb7d00ea1a7b8ebd4e68314663af',
 '5411e9269501a870cabf632f05655131',
 'bb50f2e236e5eea0100680137654686c']

In [56]:
%%javascript
$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js')

<IPython.core.display.Javascript object>